### PR TITLE
Add a ViewModel-based Controller to mobius-android

### PR DESCRIPTION
**Pull request by ZenonSeth**
This PR introduces a  class which is the equivalent of the controller (and meant to be used instead of it) but is based on the Android ViewModel.

The reasons behind introducing this:
- As an Android ViewModel, it exists between onCreate and onDestroy, allowing the Mobius Loop to run during configuration changes (e.g. rotation) and quick app minimisation/restore. The primary need for this is handling on-going network requests.
- The ViewModelController acts as a delegate for Model/State changes and View Effects so they're only handled when Lifecycle Owner is in foreground (via live data)
- Using the Live Data means that ViewEffects (e.g. navigation or show dialog) are handled properly, even if they're sent while the view is backgrounded/being re-created
- The Model and view State are kept between configuration changes since the loop is kept alive.

The concept of a View Effect (parameter V) is a type of effect that requires the corresponding Android lifecycle owner to be in an active state i.e. between onResume and onPause. 
To allow the normal effect handler to send these, the ViewModelController provides a Consumer of these View Effects to the Loop Factory Provider, which can then be passed into the normal Effect handler so it can delegate view effects where necessary

Since it's based on Android View model, this controller will keep the loop alive as long as the lifecycle owner it is associated with (via a factory to produce it) is not destroyed - meaning the Mobius loop will persist through rotations and brief app minimisation to background,

While the loop is running but the view is paused, which is between onPause and onDestroy, the controller will keep the latest model/state sent by the loop and will keep a queue of View Effects that have been sent by the effect handler. The loop is automatically disposed when the lifecycle owner is destroyed (in the ViewModel's onCleared callback)

This is a not-so-strict class data flow/interaction diagram.
Screenshot 2020-01-16 at 14 03 58

### CODE REVIEW

File: mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ * Effects that have been sent by the effect handler. The loop is automatically disposed when the
+ * lifecycle owner is destroyed
+ *
+ * @param <M> The Model with which the Mobius Loop will run
+ * @param <E> The Event type accepted by the loop
+ * @param <F> The Effect type handled by the loop
+ * @param <S> The View State which will be emitted by this controller
+ * @param <V> The View Effect which will be emitted by this controller
+ */
+public class ViewModelController<M, E, F, S, V> extends ViewModel {
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java]: I think using controller in this might be a bit confusing, especially since it has both "model" and "view" in it too. How about trying to come up with a name that makes it a bit more explicit that this is a utility for using a mobius loop as an AAC ViewModel? Maybe something along the lines of `MobiusLoopViewModel` or `MobiusAndroidViewModel`?

File: mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    super.onCleared();
+    loop.dispose();
+  }
+
+  private void onModelChanged(M model) {
+    stateData.postValue(modelToStateMapper.apply(model));
+  }
+
+  private void acceptViewEffect(V viewEffect) {
+    final Accumulator<V> currentEffects = viewEffectData.getValue();
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java]: So the idea is that you accumulate effects into the LiveData<Accumulator<>> if nobody is listening to them? it would seem reasonable to make this thread safe, it would seem to me that currently two threads could post effects at the same time and one might go missing.

I also wonder if always accumulating all effects is what you want to do? Some view effects can be quite transient, like triggering an animation, and if there is no view to observe it directly, you might want it to rather not happen at all. Maybe it'd be a good idea to make it possible to specify what behavior you want for the effects that you post?

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroid.java [hunk 3]
```diff
@@ -40,4 +42,13 @@ private MobiusAndroid() {
       MobiusLoop.Factory<M, E, F> loopFactory, M defaultModel, Init<M, F> init) {
     return Mobius.controller(loopFactory, defaultModel, init, MainThreadWorkRunner.create());
   }
+
+  @Nonnull
+  public static <V> Observer<Accumulator<V>> unfold(@Nonnull Observer<V> observer) {
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroid.java]: it would seem to me that this should rather be a method on the accumulator? I think it also would be helpful to use a name like `Accumulator.handleAll()`

Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroid.java]: I'm not sure how to fit this inside the Accumulator. 
This is just a bit of utility that might not be necessary for when subscribing to the accumulated live data, so instead of this:
```
viewModelController
  .viewEffectEmitter()
  .observe(
    lifecycleOwner, 
    accumulator -> { accumulator.handle( this::handleViewEffect )}
  )
```
we could do this
```
viewModelController
  .viewEffectEmitter()
  .observe(lifecycleOwner, unfold(this::handleViewEffect)}
```
I can always remove it since it's just utility.

File: mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        eventConsumer.accept(event);
+      }
+    }
+  }
+
+  @Nonnull
+  Accumulator<T> plus(@Nonnull T event) {
+    if (handled.compareAndSet(false, true)) {
+      events.add(event);
+      return new Accumulator<>(events);
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java]: I'm not sure I understand this combination of mutable events, and creating new accumulators that hold on to them? It sort of looks like you are trying to make an immutable linked list, that strips away handled events, but then you still mutate that list and pass it along to a new Accumulator?

Btw, it looks like the list is always a singletonList (which is immutable), so I'm pretty sure `add` always would throw `UnsupportedOperationException`.

File: mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import static java.util.Collections.singletonList;
+
+import com.spotify.mobius.functions.Consumer;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class Accumulator<T> {
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java]: This is based on immutable (-ish, because of `handled`), but couldn't this approach make it easy to accidentally create two parallel lists, where one is missing items? I think what we really want to accumulate is a list that is guaranteed to contain all items you add to it, so maybe something less FP-style would make it easier to implement and use this class correctly?

File: mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java (hunk 1 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java]: I'm happy to rename - the reason it's called a Controller is because it is meant to be used instead of a regular Controller. 

I know the regular controller doesn't have the view state in it, but that might be solvable by making the LiveData we expose publish Model udpates, and let whoever subscribes to them do their own mapping between Model->ViewState

**Comment by togi** [comment]
Maybe it's just because it's experimental, but this code really needs some tests added to it, I think there are currently some concurrency issues that aren't explored fully, and in fact I'm not sure Accumulator works at all because of `singletonList`

(As a side note, we generally try to develop most of mobius using a test-driven approach which we've found very helpful both when designing APIs as well as when dealing with some of the trickier concurrency cases)

File: mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java (hunk 2 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/ViewModelController.java]: That's a good point, I'll add the thread safety.

Accumulating and then re-playing effects seemed like the most obvious choice - I think it should be possible to specify adding an effect which won't be queued unless there's something observing it - like you said a Transient effect. I'll see about adding that too.

File: mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java (hunk 5 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java]: Yes, it is based on immutable list of events at least - and yes I can modify it to add to the current list in a more thread-safe manner too, most likely. I'll have a look to improve it.

File: mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java (hunk 4 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/Accumulator.java]: Whoops, I'll fix the signletonList call!

The list only gets mutated - but only after marking it as handled - if it hasn't been handled (it's a all-or-nothing handled flag). 
I could allocate a new list but as you mentioned if I make the whole class less FP-oriented, this would look a bit cleaner. Will try improving it.

File: mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import android.arch.lifecycle.Observer;
+
+/**
+ * An interface for an object that acts like an Android LiveData, except it will only emit each
+ * object it receives once and only once.<br>
+ * This is meant to be used as a means of sending single-handle Events to the view
+ *
+ * @param <T> The type of object to store
+ */
+public interface EffectLiveData<T> {
```
Comment (pettermahlen) [inline mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java]: This interface lacks the `removeObserver(Observer<? super T>)` method from `LiveData`. It seems like that would be useful.

File: mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius.android;
+
+import android.arch.lifecycle.LifecycleOwner;
+import android.arch.lifecycle.Observer;
+
+/**
+ * An interface for an object that acts like an Android LiveData, except it will only emit each
```
Comment (pettermahlen) [inline mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java]: This interface as defined is really only about observing. So it's more like a `LifecycleAwareObservable` or something along those lines. The main method missing (apart from one observable-related mentioned below) from `LiveData` is `getValue()`. I don't think, given the intended use of this type, that we should add that. 

The semantics around queueing/replaying T:s are defined in an implementation as things stand today. I'm not sure if that should somehow be reflected in the naming and documentation. I guess the only reason for introducing this interface is to have a place to define the queuing/replaying semantics? I think we should give the name of this some more thought. Some ideas:

- QueuingLiveData
- LifecycleAwareObservable
- ViewEffectQueue (if it's made very un-generic and only supports the single use-case we use it for today).

File: mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java (hunk 6 above)
Comment (pettermahlen) [reply mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java]: Or maybe it should be removed, according to the "when in doubt, leave it out" principle (see talk by Josh Bloch or https://dl.acm.org/doi/10.1145/1176617.1176622).

File: mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    data.observeForever(unfold(observer));
+  }
+
+  @Override
+  public void removeObservers(LifecycleOwner owner) {
+    data.removeObservers(owner);
+  }
+
+  /**
+   * This method will post the value via the live thread, behaving similarly to MutableLiveData's
```
Comment (pettermahlen) [inline mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java]: What is the 'live thread'?

File: mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+   * postValue.<br>
+   * However, if nothing is observing the data at the moment, the value is simply discarded.
+   *
+   * @param value The value to emit, which will be discarded if nothing is observing it
+   */
+  void postTransient(T value) {
```
Comment (pettermahlen) [inline mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java]: Are we sure this method is necessary?

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+ */
+package com.spotify.mobius.android;
+
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.ViewModel;
+import com.spotify.mobius.First;
+import com.spotify.mobius.Init;
+import com.spotify.mobius.MobiusLoop;
+import com.spotify.mobius.MobiusLoop.Factory;
```
Comment (pettermahlen) [inline mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java]: nit: I think the code is more readable if this isn't statically imported - the `MobiusLoop.` part of the name is useful to see which kind of factory it is.

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+ * and onPause. To allow the normal effect handler to send these, the controller will provide a
+ * Consumer of these View Effects to the Loop Factory Provider, which can then be passed into the
+ * normal Effect handler so it can delegate view effects where necessary<br>
+ *
+ * <p>Since it's based on Android View model, this controller will keep the loop alive as long as
+ * the lifecycle owner it is associated with (via a factory to produce it) is not destroyed -
+ * meaning the Mobius loop will persist through rotations and brief app minimization to background,
+ * <br>
+ *
+ * <p>While the loop is running but the view is paused, which is between onPause and onDestroy, the
```
[the remark this answers is not in the document]
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java]: Yes, I need to fix this, though the Android ViewModel runs until on-destroy and it's starting time obviously depends on where it's initialised - which usually should be in onCreate or onCreateView (how Mobius is currently used in Spotify in the Kids app)

File: mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java (hunk 6 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java]: The reason I haven't included the `removeObserver` call is that we wrap the given observers into other observers, so it wouldn't be a direct pass-through call to the MutableLiveData, but we would have to somehow keep a map. Keeping a map then means we'll have to worry about the lifecycle of objects (though possibly WeakReference would be enough) - and I didn't want to add that complication into this - which is why I excluded the `removeObserver` call

File: mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java (hunk 8 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java]: That should have been `main thread`, will fix

File: mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java (hunk 9 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java]: In my initial PR this wasn't included. This came about as discussion with @togi who suggested a possible use for it - and since it was fairly easy to implement I've added it.

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java (hunk 10 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java]: 👍

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+ * Effects that have been sent by the effect handler. The loop is automatically disposed when the
+ * lifecycle owner is destroyed
+ *
+ * @param <M> The Model with which the Mobius Loop will run
+ * @param <E> The Event type accepted by the loop
+ * @param <F> The Effect type handled by the loop
+ * @param <S> The View State which will be emitted by this controller
+ * @param <V> The View Effect which will be emitted by this controller
+ */
+public class MobiusAndroidViewModel<M, E, F, S, V> extends ViewModel {
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java]: I thought some more about this name - maybe it'd be a good idea to call it MobiusLoopViewModel after all, to avoid confusing it with the AndroidViewModel class from AAC (which has nothing to do with this class)

File: mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+ * object it receives once and only once.<br>
+ * This is meant to be used as a means of sending single-handle Events to the view
```
[the remark this answers is not in the document]
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java]: Effects - will fix!

File: mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java (hunk 7 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/EffectLiveData.java]: The reason I had to write this interface is to expose an observable-only live-data like object (basically hide the post/postTransient calls) - and unfortunately, the way Google has defined Live Data it's an abstract class, not an interface, so I couldn't use that. 

Yea, this shouldn't have getValue since that would then mean exposing the Accumulator, which the rest of the work here tried not to do. I'll rename it to something better 👍

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java (hunk 12 above)
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java]: 👍 - will rename to that, seems good

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    M model = loop.getMostRecentModel();
+    return model != null ? model : startModel;
+  }
+
+  @Nonnull
+  public final LiveData<S> stateEmitter() {
+    return stateData;
+  }
+
+  public final EffectLiveData<V> viewEffectEmitter() {
```
[the remark this answers is not in the document]
Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MobiusAndroidViewModel.java]: 👍

File: mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+   * observing the data, they will be queued up along with any other calls made to post values.
+   *
+   * @param value The value to emit, or alternatively if nothing is observing, to queue up
+   */
+  void post(T value) {
+    synchronized (data) {
+      if (data.getValue() == null) {
+        data.postValue(new Accumulator<T>().append(value));
+      } else {
+        data.postValue(data.getValue().append(value));
```
Comment (pettermahlen) [inline mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java]: I guess it's not really necessary to call `data.postValue()` here. Maybe this method could be simplified to something like:

```
  synchronized(data) {
    Accumulator<T> acc = data.getValue();
    if (acc == null) {
       acc = new Accumulator<>();
       data.postValue(acc);
    }
    acc.append(value);
  }
```
A reason is that `data.postValue` is scary since it might miss things that are not posted on the main thread. In general, perhaps using a LiveData isn't the safest thing to do.

Comment (ZenonSeth) [reply mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java]: `data.postValue()` is necessary.

First, the `getValue()` call will only be null once, when there's nothing posted at all. Afterwards, the object posted will be then received and will internally use up its values. 

When we call post for a second time, we will append new values to the object, but unless we posted it again into the MutableLiveData, whoever is observing the live data won't get notified that we've sent out new things.

Ideally (and originally) I had a more of FP approach where instead of append there was a function called Plus that always returned an new Accumulator, but since then for optimizations of allocations I've converted it to a single mutable object.

Bottom line though, we definitely need to call postValue every time, otherwise observers won't be notified by the live data that there's new things to observe

File: mobius-android/src/main/java/com/spotify/mobius/android/LiveQueue.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+   *     Paused and to resume sending effects once it resumes.
+   * @param liveEffectsObserver This required observer will be forwarded all effects while the
+   *     lifecycle owner is in a Resumed state.
+   * @param pausedEffectsObserver The nullable observer will be invoked when the lifecycle owner
+   *     resumes, and will receive a queue of effects, ordered as they occurred while paused.
+   */
+  void setObserver(
+      @Nonnull LifecycleOwner lifecycleOwner,
+      @Nonnull Observer<? super T> liveEffectsObserver,
+      @Nullable Observer<Queue<? super T>> pausedEffectsObserver);
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/LiveQueue.java]: I'd say let's just expose this as an Iterable instead of a Queue, I don't think there is any particular point to exposing the "queueness" of it.

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusLoopViewModel.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  @Nonnull
+  public final M getModel() {
+    M model = loop.getMostRecentModel();
+    return model != null ? model : startModel;
+  }
+
+  @Nonnull
+  public final LiveData<M> stateEmitter() {
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/MobiusLoopViewModel.java]: `stateEmitter` -> `getModel`/`getModels`/something that would be a typical name for a LiveData getter?

File: mobius-android/src/main/java/com/spotify/mobius/android/MobiusLoopViewModel.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    return modelData;
+  }
+
+  public final LiveQueue<V> viewEffectEmitter() {
```
Comment (togi) [inline mobius-android/src/main/java/com/spotify/mobius/android/MobiusLoopViewModel.java]: `viewEffectEmitter` -> `getViewEffects`/something that matches the naming schema of the previous getter?

File: mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+    singleLiveData.post("two");
+    fakeLifecycleOwner2.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
+    singleLiveData.setObserver(fakeLifecycleOwner2, liveObserver, pausedObserver);
+
+    assertThat(liveObserver.valueCount(), equalTo(0));
+    pausedObserver.assertValues(queueOf("one", "two"));
+  }
+
+  @Test
+  public void shouldSendQueuedEffectsIfObserverSwitchedWithoutClearing() {
```
Comment (pettermahlen) [inline mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java]: Is 'clearing' correct as a test description here? It looks more like it's about whether the observer is active or not before or after adding?

Comment (pettermahlen) [reply mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java]: (same for the above test)

File: mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+import com.spotify.mobius.runners.WorkRunners;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.Queue;
+import org.junit.Before;
+import org.junit.Test;
+
+public class MutableLiveQueueTest {
+
+  private MutableLiveQueue<String> singleLiveData;
```
Comment (pettermahlen) [inline mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java]: This field should be renamed to `liveQueue`

File: mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java (hunk 19 above)
Comment (pettermahlen) [reply mobius-android/src/test/java/com/spotify/mobius/android/MutableLiveQueueTest.java]: I didn't read it correctly, it's fine.
